### PR TITLE
feat: use convict for env vars

### PR DESCRIPTION
**Pull request by harishv7**
## Problem

We should have an exhaustive env variables check at service startup. This ensures that we can confidently use environment variables across our codebase without worrying about situations where the expected env var is missing or empty.

## Solution

We can use the popular convict npm package which helps us create a schema of expected env vars and assert that these are present at system startup.

This  introduces a configuration schema that ensure the environment is predictable and maintainable.

Do help to verify if all the env vars are captured well in the config file. There is one variable which me and @alexanderleegs were not able to identify the usage: `AWS_BACKEND_EB_ENV_NAME` used in`logger.js`. This variable was not existing on prod EB env. When added to staging, it didn't make a change to the staging log-groups produced as well.

This has been excluded in the config. But we should decided if this variable is required or not.

**Breaking Changes**

- [ ] Yes - this PR contains breaking changes
  - Details ...
- [ ] No - this PR is backwards compatible

## Tests

Do run `npm run test --runInBand` to ensure all the tests are passing

## Deploy Notes

Ensure that all required variables are present in the environment files.
In the absence of the variables in `.env` the defaults will be used.

**New environment variables**:

Please check `config.ts` under `src/config` to ensure all the env variables match

**New dependencies**:

- `"convict": "^6.2.4",` : introduces a configuration schema that ensure the environment is predictable and maintainable

**New dev dependencies**:

- `@types/convict` : types for convict package

### CODE REVIEW

File: src/config/config.ts [hunk 1]
```diff
@@ -0,0 +1,360 @@
...
+      format: "required-string",
+      default: "",
+    },
+  },
+  cypress: {
+    e2eTestRepo: {
+      doc: "Name of the e2e test GitHub repository",
+      env: "E2E_TEST_REPO",
+      format: "required-string",
+      default: null,
```
Comment (seaerchin) [inline src/config/config.ts:308]: can we just give the actual e2e repo name here cos idt it's sensitive

File: src/config/config.ts [hunk 2]
```diff
@@ -0,0 +1,360 @@
...
+      format: "required-boolean",
+      default: false,
+    },
+  },
+})
+
+// Perform validation
+config.validate({ allowed: "strict" })
+
+module.exports = config
```
Comment (seaerchin) [inline src/config/config.ts:358]: wa why mix commonJS exports and es2021 exports :(

**Review by seaerchin — COMMENTED** [review]
minor comments; on a separate note, do we need to define the `required-string/number` format cos `convict` uses the prototype to [validate](https://github.com/mozilla/node-convict/tree/master/packages/convict#var-config--convictschema-opts), which rejects `null/undefined`

**Comment by seaerchin** [comment]
also seems like got failing test; could i trouble you to look at it?

File: src/config/config.ts [hunk 3]
```diff
@@ -0,0 +1,360 @@
...
+  },
+})
+
+convict.addFormat({
+  name: "required-number",
+  validate: (val: any) => {
+    if (!val) throw new Error("value cannot be empty, null or undefined")
+    if (typeof val !== "number") throw new Error("value must be a number")
+  },
+  coerce: (val: string) => {
```
[the remark this answers is not in the document]
Comment (harishv7) [reply src/config/config.ts:17]: Actually if you console log and check it runs coerce first then validate
Screenshot 2023-03-07 at 1 44 29 PM

File: src/config/config.ts [hunk 4]
```diff
@@ -0,0 +1,360 @@
...
+    if (!val) throw new Error("value cannot be empty, null or undefined")
+    if (typeof val !== "string") throw new Error("value must be a string")
+  },
+})
+
+convict.addFormat({
+  name: "required-number",
+  validate: (val: any) => {
+    if (!val) throw new Error("value cannot be empty, null or undefined")
+    if (typeof val !== "number") throw new Error("value must be a number")
```
[the remark this answers is not in the document]
Comment (harishv7) [reply src/config/config.ts:15]: It will run coerce first, so validate will be the latter step

File: src/config/config.ts [hunk 5]
```diff
@@ -0,0 +1,360 @@
...
+  },
+  coerce: (val: string) => String(val).toLowerCase() === "true",
+})
+
+// Define a schema
+const config = convict({
+  env: {
+    doc: "The application environment.",
+    env: "NODE_ENV",
+    format: "required-string",
```
[the remark this answers is not in the document]
Comment (harishv7) [reply src/config/config.ts:41]: agreed, done

File: src/config/config.ts [hunk 6]
```diff
@@ -0,0 +1,360 @@
...
+convict.addFormat({
+  name: "required-string",
+  validate: (val: any) => {
+    if (!val) throw new Error("value cannot be empty, null or undefined")
+    if (typeof val !== "string") throw new Error("value must be a string")
+  },
+})
+
+convict.addFormat({
+  name: "required-number",
```
[the remark this answers is not in the document]
Comment (harishv7) [reply src/config/config.ts:12]: done

File: .eslintrc.json [hunk 7]
```diff
@@ -30,6 +30,12 @@
                 "newlines-between": "always",
                 "groups": ["builtin", "external", "internal", "parent", "sibling", "index", "object"],
                 "pathGroups": [{
+                    "pattern": "@config/**",
+                    "group": "internal",
+                    "position": "before"
+                    },
+                    {
+
```
[the remark this answers is not in the document]
Comment (harishv7) [reply .eslintrc.json:38]: done, somehow prettier is configured to ignore this file so it didnt format when I committed

File: src/config/config.ts [hunk 8]
```diff
@@ -0,0 +1,360 @@
...
+      format: "required-string",
+      default: null,
+    },
+  },
+  auth: {
+    cookieDomain: {
+      doc: "Domain to set for auth cookie",
+      env: "COOKIE_DOMAIN",
+      format: "required-string",
+      default: "localhost",
```
[the remark this answers is not in the document]
Comment (harishv7) [reply src/config/config.ts:98]: @seaerchin how can I know the exhaustive list?

File: src/config/config.ts (hunk 2 above)
Comment (harishv7) [reply src/config/config.ts:355]: Hmm yea, I was trying use `import` syntax everywhere, but didnt work, had to use require in some places

File: src/config/config.ts [hunk 9]
```diff
@@ -0,0 +1,360 @@
...
+      doc: "GitHub organization that owns all site repositories",
+      env: "GITHUB_ORG_NAME",
+      format: "required-string",
+      default: "isomerpages",
+    },
+    buildOrgName: {
+      doc: "GitHub organization that owns the build repository",
+      env: "GITHUB_BUILD_ORG_NAME",
+      format: "required-string",
+      default: "opengovsg",
```
[the remark this answers is not in the document]
Comment (harishv7) [reply src/config/config.ts:215]: i checked but we are using `opengovsg` currently

File: src/config/config.ts (hunk 1 above)
Comment (harishv7) [reply src/config/config.ts:305]: @seaerchin updated to `"e2e-test-repo"`

File: src/config/config.ts [hunk 10]
```diff
@@ -0,0 +1,360 @@
...
+      default: null,
+    },
+  },
+  database: {
+    dbUri: {
+      doc: "Database URI",
+      env: "DB_URI",
+      sensitive: true,
+      format: "required-string",
+      default: null,
```
[the remark this answers is not in the document]
Comment (harishv7) [reply src/config/config.ts:329]: done

File: src/config/config.ts [hunk 11]
```diff
@@ -0,0 +1,357 @@
...
+      sensitive: true,
+      format: "required-string",
+      default: "",
+    },
+    domainWhitelist: {
+      doc: "List of domains that are allowed to access the API",
+      env: "DOMAIN_WHITELIST",
+      format: "required-string",
+      default: null,
+    },
```
Comment (alexanderleegs) [inline src/config/config.ts:144]: Also unused env var!

File: src/services/identity/__tests__/SitesService.spec.ts [hunk 12]
```diff
@@ -302,8 +304,8 @@ describe("SitesService", () => {
   describe("getSites", () => {
     it("Filters accessible sites for github user correctly", async () => {
       // Store the API key and set it later so that other tests are not affected
-      const currRepoCount = process.env.ISOMERPAGES_REPO_PAGE_COUNT
-      process.env.ISOMERPAGES_REPO_PAGE_COUNT = "3"
+      const currRepoCount = config.get("sites.pageCount")
```
[the remark this answers is not in the document]
Comment (harishv7) [reply src/services/identity/__tests__/SitesService.spec.ts:307]: We are getting and immediately setting it, so I thought it can be placed together for clarity as they go together as a "pair". Alternatively, we can define a single `currRepoCount` at the top and then only do the `set` in here.
```
const currRepoCount = config.get("sites.pageCount")
config.set("sites.pageCount", 3)
```

File: src/config/config.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (alexanderleegs) [reply src/config/config.ts:215]: Ah the build repo `isomer-build` is under the `opengovsg` org, not `isomerpages`, so this value is correct

File: src/database/config.js [hunk 13]
```diff
@@ -1,10 +1,14 @@
...
 // TODO: This came from a past project - I don't remember why I wrote this but let's explore later.
 // We have to manually parse database URL because sequelize-typescript requires explicit
 // connection parameters.
 
-const { DB_URI, DB_MIN_POOL, DB_MAX_POOL } = process.env
+const DB_URI = config.get("database.dbUri")
+const DB_MIN_POOL = config.get("database.dbMinPool")
+const DB_MAX_POOL = config.get("database.dbMaxPool")
 
 if (!DB_URI) throw new Error("DB_URI is not defined")
```
Comment (kishore03109) [inline src/database/config.js:13]: chore: check not needed anymore

File: src/database/config.js [hunk 14]
```diff
@@ -1,10 +1,14 @@
...
 const parsed = parse(DB_URI)
```
Comment (kishore03109) [inline src/database/config.js:14]: chore: parsing of `DB_URI` not needed anymore

File: src/routes/formsgSiteCreation.ts [hunk 15]
```diff
@@ -14,7 +16,7 @@ import UsersService from "@services/identity/UsersService"
 import InfraService from "@services/infra/InfraService"
 import { mailer } from "@services/utilServices/MailClient"
 
-const { SITE_CREATE_FORM_KEY } = process.env
+const SITE_CREATE_FORM_KEY = config.get("formSg.siteCreateFormKey")
```
Comment (kishore03109) [inline src/routes/formsgSiteCreation.ts:19]: chore: no need to validate existence of key in line 142

File: src/bootstrap/index.ts [hunk 16]
```diff
@@ -80,7 +80,7 @@ const createListener = (
 }
 
 const bootstrapServer = (app: Express) => {
-  const port = normalizePort(process.env.PORT || "8081") // TODO: PORT is not in env example
+  const port = normalizePort(process.env.PORT || "8081")
```
Comment (kishore03109) [inline src/bootstrap/index.ts:83]: hmm is there a reason why we dont use convict for this?

File: src/config/config.ts [hunk 17]
```diff
@@ -1,7 +1,4 @@
 import convict from "convict"
-import * as dotenv from "dotenv"
-
-dotenv.config()
```
Comment (kishore03109) [inline src/config/config.ts]: Just to check in, was this deletion of code intended?

**Review by kishore03109 — CHANGES_REQUESTED** [review]
Chore: I think there are some parts of the code where there were some parsing logic for the env variables, which is now done by convict, and thus are no longer needed. (I got lazy to search through them towards the end HAHA)
 
Style: I think its cleaner to do a rebase rather than a merge to keep the commit changes clean, making the code easier to review! Let me know if you dont prefer this workflow!

File: src/config/config.ts (hunk 2 above)
Comment (seaerchin) [reply src/config/config.ts:355]: could i get more insight into this? it sounds p important

File: src/config/config.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (seaerchin) [reply src/config/config.ts:98]: think can find in the env var for the respective env

File: src/config/config.ts (hunk 17 above)
Comment (harishv7) [reply src/config/config.ts]: yea, convict does the env load automatically, we dont need dotenv

File: src/config/config.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (harishv7) [reply src/config/config.ts:98]: ok added

File: src/config/config.ts [hunk 18]
```diff
@@ -0,0 +1,357 @@
...
+      format: "required-number",
+      default: 10,
+    },
+    localSiteAccessToken: {
+      doc: "Access token for local site",
+      env: "LOCAL_SITE_ACCESS_TOKEN",
+      sensitive: true,
+      format: "required-string",
+      default: null,
+    },
```
[the remark this answers is not in the document]
Comment (harishv7) [reply src/config/config.ts:91]: okay removed

File: src/config/config.ts [hunk 19]
```diff
@@ -0,0 +1,357 @@
...
+      format: "required-number",
+      default: 900000,
+    },
+    otpSecret: {
+      doc: "Secret used for OTP generation",
+      env: "OTP_SECRET",
+      sensitive: true,
+      format: "required-string",
+      default: "",
+    },
```
[the remark this answers is not in the document]
Comment (harishv7) [reply src/config/config.ts:138]: Yea I added a TODO, have added an issue here https://github.com/isomerpages/isomercms-backend/issues/648

File: src/config/config.ts (hunk 11 above)
Comment (harishv7) [reply src/config/config.ts]: removed

File: src/services/identity/index.ts [hunk 20]
```diff
@@ -21,16 +23,13 @@ import NotificationsService from "./NotificationsService"
...
-if (!OTP_SECRET) {
-  throw new Error(
-    "Please ensure that you have set OTP_SECRET in your env vars and that you have sourced them!"
-  )
-}
-
+// TODO: To remove TOTP
 const totpGenerator = new TotpGenerator({
   secret: OTP_SECRET!,
   expiry: parseInt(OTP_EXPIRY!, 10) ?? undefined,
```
[the remark this answers is not in the document]
Comment (harishv7) [reply src/services/identity/index.ts]: outdated

File: src/config/config.ts [hunk 21]
```diff
@@ -0,0 +1,341 @@
...
+      format: "required-boolean",
+      default: "false",
+    },
+  },
+})
+
+// Perform validation
+config.validate({ allowed: "strict" })
+
+// console.log("env", config.get("env"))
```
[the remark this answers is not in the document]
Comment (harishv7) [reply src/config/config.ts]: outdated

File: config.js [hunk 22]
```diff
@@ -0,0 +1,208 @@
+const convict = require("convict")
+
+const { NODE_ENV } = process.env
+const isLocalDev = NODE_ENV === "LOCAL_DEV"
+
+// validate in all environments
+convict.addFormat({
+  name: "required-string",
```
[the remark this answers is not in the document]
Comment (harishv7) [reply config.js]: as discussed, we wanted to define our own validations for extensibility and for special checks

File: src/bootstrap/index.ts (hunk 16 above)
Comment (harishv7) [reply src/bootstrap/index.ts]: fixed

File: config.js (hunk 22 above)
[the remark this answers is not in the document]
Comment (kishore03109) [reply config.js]: as discussed offline, we are using this to ensure empty strings still throw an error

File: src/database/config.js (hunk 14 above)
Comment (harishv7) [reply src/database/config.js:14]: This parse is to extract the port from a connection string. it will be still needed

**Comment by harishv7** [comment]
> Chore: I think there are some parts of the code where there were some parsing logic for the env variables, which is now done by convict, and thus are no longer needed. (I got lazy to search through them towards the end HAHA)
> 
> Style: I think its cleaner to do a rebase rather than a merge to keep the commit changes clean, making the code easier to review! Let me know if you dont prefer this workflow!

- Okay will go through remove those checks
- Rebase or merge, we still do a squash and merge eventually right? Does that matter on a PR level?

File: src/services/identity/__tests__/SitesService.spec.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (harishv7) [reply src/services/identity/__tests__/SitesService.spec.ts:307]: @seaerchin Example if you change it to use `import` syntax in `src/services/fileServices/MdPageServices/MediaFileService.js`:

```
import config from "@config/config"

```

you will get the following error when running tests:

```
 FAIL  src/services/fileServices/MdPageServices/__tests__/MediaFileService.spec.js
  ● Test suite failed to run

    Jest encountered an unexpected token

    Jest failed to parse a file. This happens e.g. when your code or its dependencies use non-standard JavaScript syntax, or when Jest is not configured to support such syntax.

    Out of the box Jest supports Babel, which will be used to transform your files into valid JS based on your Babel configuration.

    By default "node_modules" folder is ignored by transformers.

    Here's what you can do:
     • If you are trying to use ECMAScript Modules, see https://jestjs.io/docs/ecmascript-modules for how to enable it.
     • If you are trying to use TypeScript, see https://jestjs.io/docs/getting-started#using-typescript
     • To have some of your "node_modules" files transformed, you can specify a custom "transformIgnorePatterns" in your config.
     • If you need a custom transformation specify a "transform" option in your config.
     • If you simply want to mock your non-JS modules (e.g. binary assets) you can stub them out with the "moduleNameMapper" config option.

    You'll find more details and examples of these config options in the docs:
    https://jestjs.io/docs/configuration
    For information about custom transformations, see:
    https://jestjs.io/docs/code-transformation

    Details:

    /Users/harishv/Documents/Code/isomer/isomercms-backend/src/services/fileServices/MdPageServices/MediaFileService.js:1
    ({"Object.<anonymous>":function(module,exports,require,__dirname,__filename,jest){import config from "@config/config"
                                                                                      ^^^^^^

    SyntaxError: Cannot use import statement outside a module

      39 |   const {
      40 |     MediaFileService,
    > 41 |   } = require("@services/fileServices/MdPageServices/MediaFileService")
         |       ^
      42 |   const service = new MediaFileService({
      43 |     gitHubService: mockGithubService,
      44 |   })

      at Runtime.createScriptFromCode (../node_modules/jest-runtime/build/index.js:1728:14)
      at services/fileServices/MdPageServices/__tests__/MediaFileService.spec.js:41:7
      at Object.<anonymous> (services/fileServices/MdPageServices/__tests__/MediaFileService.spec.js:7:1)
```

File: src/config/config.ts (hunk 2 above)
Comment (harishv7) [reply src/config/config.ts:349]: @seaerchin oops replied in below thread
https://github.com/isomerpages/isomercms-backend/pull/646#discussion_r1127616701

Comment (seaerchin) [reply src/config/config.ts:349]: it works if we use named exports rather than default ._. i've a feeling this is because the transpilation doesn't map [cleanly](https://stackoverflow.com/questions/40294870/module-exports-vs-export-default-in-node-js-and-es6)...

if you don't mind, i can commit directly here

File: .eslintrc.json
Comment (seaerchin) [inline .eslintrc.json:1]: did you actually modify this?>>

File: src/config/config.ts [hunk 23]
```diff
@@ -0,0 +1,351 @@
...
+  },
+  coerce: (val: string) => String(val).toLowerCase() === "true",
+})
+
+// Define a schema
+const config = convict({
+  env: {
+    doc: "The application environment.",
+    env: "NODE_ENV",
+    format: ["LOCAL_DEV", "DEV", "test", "prod"],
```
Comment (seaerchin) [inline src/config/config.ts:41]: need to add staging also, no? we might also use this chance to eliminate `local_dev` (since we don't have a diff eb setup separately for dev)

File: src/config/config.ts [hunk 24]
```diff
@@ -0,0 +1,351 @@
...
+      sensitive: true,
+      format: "required-string",
+      default: null,
+    },
+  },
+  auth: {
+    cookieDomain: {
+      doc: "Domain to set for auth cookie",
+      env: "COOKIE_DOMAIN",
+      format: ["localhost", "cms.isomer.gov.sg", "isomer.gov.sg"],
```
Comment (seaerchin) [inline src/config/config.ts:97]: could i ask if this also includes the staging cookie domain?

**Review by seaerchin — COMMENTED** [review]
minor comments, not approving first for safety

File: src/config/config.ts (hunk 23 above)
Comment (harishv7) [reply src/config/config.ts:41]: our staging is using "DEV" as the value. But I will add in another option

File: .eslintrc.json
Comment (harishv7) [reply .eslintrc.json:1]: I didn't modify, but I formatted it with prettier

File: src/config/config.ts (hunk 23 above)
Comment (seaerchin) [reply src/config/config.ts:41]: dun add, we shud eliminate the `local_dev` option cos it's useless HAHA

File: src/config/config.ts (hunk 24 above)
Comment (harishv7) [reply src/config/config.ts:97]: I checked aws console for staging EB and i included that as well. let me know if there should be others i should add

File: src/config/config.ts (hunk 2 above)
Comment (harishv7) [reply src/config/config.ts:349]: @seaerchin sure go ahead

File: src/config/config.ts (hunk 23 above)
Comment (harishv7) [reply src/config/config.ts]: @seaerchin i won't suggest to remove it in this PR though. We do have some logic where its being used, e.g. in `src/services/identity/index.ts` and so on

Should do this clean up separately

File: src/services/utilServices/MailClient.ts [hunk 25]
```diff
@@ -39,16 +41,11 @@ class MailClient {
 }
 export default MailClient
 
-const { NODE_ENV, POSTMAN_API_KEY } = process.env
+const NODE_ENV = config.get("env")
+const POSTMAN_API_KEY = config.get("postman.apiKey")
```
Comment (kishore03109) [inline src/services/utilServices/MailClient.ts:45]: Ok for some reason I cannot comment directly on line 58, but 
`: new MailClient(POSTMAN_API_KEY!)` can be changed to `: new MailClient(POSTMAN_API_KEY)` since we have guarentee its a `string` already

File: src/routes/formsgSiteCreation.ts [hunk 26]
```diff
@@ -137,11 +139,6 @@ export class FormsgRouter {
...
     const router = express.Router({ mergeParams: true })
 
-    if (!SITE_CREATE_FORM_KEY) {
-      throw new InitializationError(
-        "Required SITE_CREATE_FORM_KEY environment variable is empty."
-      )
-    }
     router.post(
       "/create-site",
       attachFormSGHandler(SITE_CREATE_FORM_KEY || ""),
```
Comment (kishore03109) [inline src/routes/formsgSiteCreation.ts:144]: Chore: change `attachFormSGHandler(SITE_CREATE_FORM_KEY || "")` to `attachFormSGHandler(SITE_CREATE_FORM_KEY)`

**Review by kishore03109 — COMMENTED** [review]
mostly lgtm but a few comments. 
I do have a concern due to the strictness setting that we have set for this. Currently, this will break if any one of our variables are set wrongly, causing prod to be down. Let me know your thoughts about this suggestion:

we can tackle this in 2 prs:
1) this PR, but mode is set to `WARN` rather than `STRICT`
we look at the logs, make sure its not throwing any unexpected errors. The following week when we run release
2) we change the mode from `WARN` to `STRICT`

that way we ensure that things dont break, allowing for a smoother deployment.

File: src/server.js [hunk 27]
```diff
@@ -80,10 +80,10 @@ const express = require("express")
 const helmet = require("helmet")
 const createError = require("http-errors")
 
-const NODE_ENV = config.get("env")
-
 const isSecure =
-  NODE_ENV !== "DEV" && NODE_ENV !== "LOCAL_DEV" && NODE_ENV !== "test"
+  process.env.NODE_ENV !== "DEV" &&
```
[the remark this answers is not in the document]
Comment (kishore03109) [reply src/server.js]: outdated comment opps

File: src/config/config.ts (hunk 2 above)
Comment (seaerchin) [reply src/config/config.ts]: done

File: src/config/config.ts (hunk 23 above)
Comment (seaerchin) [reply src/config/config.ts]: sure, that's fine too; can track in issue?

Comment (harishv7) [reply src/config/config.ts]: Sure, it's here: https://github.com/isomerpages/isomercms-backend/issues/656

File: src/services/utilServices/MailClient.ts (hunk 25 above)
Comment (harishv7) [reply src/services/utilServices/MailClient.ts:45]: okay done

File: src/routes/formsgSiteCreation.ts (hunk 26 above)
Comment (harishv7) [reply src/routes/formsgSiteCreation.ts:144]: good catch, done

**Comment by harishv7** [comment]
> mostly lgtm but a few comments. I do have a concern due to the strictness setting that we have set for this. Currently, this will break if any one of our variables are set wrongly, causing prod to be down. Let me know your thoughts about this suggestion:
> 
> we can tackle this in 2 prs:
> 
> 1. this PR, but mode is set to `WARN` rather than `STRICT`
>    we look at the logs, make sure its not throwing any unexpected errors. The following week when we run release
> 2. we change the mode from `WARN` to `STRICT`
> 
> that way we ensure that things dont break, allowing for a smoother deployment.

Some clarifications:
- If any vars are set wrongly or not set, we will know at build time itself since app will not startup
- Prod envs should be set correctly, hence a "hard failure" should be expected when important things like env vars are missing/wrong

Basically when deploying, we should be 100% certain what we are deploying is correct, especially with things like config/envs

Did i get your concern correctly? But if the team feels we should not have a "hard" failure, we can discuss the deploy strategy as well

cc @seaerchin @alexanderleegs @QiluXie

**Comment by seaerchin** [comment]
> > mostly lgtm but a few comments. I do have a concern due to the strictness setting that we have set for this. Currently, this will break if any one of our variables are set wrongly, causing prod to be down. Let me know your thoughts about this suggestion:
> > we can tackle this in 2 prs:
> > 
> > 1. this PR, but mode is set to `WARN` rather than `STRICT`
> >    we look at the logs, make sure its not throwing any unexpected errors. The following week when we run release
> > 2. we change the mode from `WARN` to `STRICT`
> > 
> > that way we ensure that things dont break, allowing for a smoother deployment.
> 
> Some clarifications:
> 
> * If any vars are set wrongly or not set, we will know at build time itself since app will not startup
> * Prod envs should be set correctly, hence a "hard failure" should be expected when important things like env vars are missing/wrong
> 
> Basically when deploying, we should be 100% certain what we are deploying is correct, especially with things like config/envs
> 
> Did i get your concern correctly? But if the team feels we should not have a "hard" failure, we can discuss the deploy strategy as well
> 
> cc @seaerchin @alexanderleegs @QiluXie

if `convict` fails at start up, i'm more in favour of failing fast! env vars are something that is difficult to debug/catch (even when warn), and we should try to fail at start up - that's also the reason why we do 
```ts
if (!SOME_ENV_VAR) throw Error()
```
but because this fails when the file is used, it's still not as ideal. failing fast here has 2 benefits: 
1. we can easily rollback to a working version whilst figuring out what's wrong 
2. no tricky bugs involved (exact env var will be pointed out)

**Comment by kishore03109** [comment]
@harishv7 
> If any vars are set wrongly or not set, we will know at build time itself since app will not startup

Woah am confused about this since my understanding is that it will fail at runtime rather than build time, but ill clarify with you offline instead haha

> Prod envs should be set correctly, hence a "hard failure" should be expected when important things like env vars are missing/wrong 

Oh yes, agreed that this should be the end state. My concern primarily stemmed from the fact we are merging into develop, and develop and identity branch has diverged to some extent, I was worried that the merge itself will have breaking changes. But, I agree with @seaerchin about failing fast since we do have rolling deployments anyways, so prod wont be down even if the app fails at runtime. 

TLDR: lgtm

**Review by kishore03109 — APPROVED** [review]

**Comment by harishv7** [comment]
> > @harishv7
> > > If any vars are set wrongly or not set, we will know at build time itself since app will not startup
> > 
> > 
> > Woah am confused about this since my understanding is that it will fail at runtime rather than build time, but ill clarify with you offline instead haha
> > > Prod envs should be set correctly, hence a "hard failure" should be expected when important things like env vars are missing/wrong
> > 
> > 
> > Oh yes, agreed that this should be the end state. My concern primarily stemmed from the fact we are merging into develop, and develop and identity branch has diverged to some extent, I was worried that the merge itself will have breaking changes. But, I agree with @seaerchin about failing fast since we do have rolling deployments anyways, so prod wont be down even if the app fails at runtime.
> > TLDR: lgtm
> 
> @kishore03109 Yea, you're right, it is technically runtime, but at service startup.
> 
> @kishore03109 @seaerchin @alexanderleegs We are actually merging into `vapt` branch first. Is that the right destination branch? For deploy strategy, we need to deploy this on staging first. What would be the best way to do that?

Okay confirmed destination branch should be `identity-phase-2` with @alexanderleegs
